// fusion-plugin-react-router/dist-browser-esm/index.js
/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */
import plugin, { RouterProviderToken, RouterToken, GetStaticContextToken } from './plugin';
import * as browser from './browser';
export default plugin;
export { RouterProviderToken, RouterToken, GetStaticContextToken };
export const BrowserRouter = true ? browser.BrowserRouter : server.BrowserRouter;
export const HashRouter = true ? browser.HashRouter : server.HashRouter;
export const Link = true ? browser.Link : server.Link;
export const matchPath = true ? browser.matchPath : server.matchPath;
export const MemoryRouter = true ? browser.MemoryRouter : server.MemoryRouter;
export const NavLink = true ? browser.NavLink : server.NavLink;
export const Prompt = true ? browser.Prompt : server.Prompt;
export const Route = true ? browser.Route : server.Route;
export const Router = true ? browser.Router : server.Router;
export const Switch = true ? browser.Switch : server.Switch;
export const withRouter = true ? browser.withRouter : server.withRouter;
export const NotFound = true ? browser.NotFound : server.NotFound;
export const Redirect = true ? browser.Redirect : server.Redirect;
export const Status = true ? browser.Status : server.Status;
export const useHistory = true ? browser.useHistory : server.useHistory;
export const useRouteMatch = true ? browser.useRouteMatch : server.useRouteMatch;
export const useLocation = true ? browser.useLocation : server.useLocation;
export const useParams = true ? browser.useParams : server.useParams;
