// fusion-plugin-react-router/dist-browser-cjs/server.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "_Router", {
  enumerable: true,
  get: function () {
    return _reactRouterDom.Router;
  }
});
Object.defineProperty(exports, "Status", {
  enumerable: true,
  get: function () {
    return _Status.Status;
  }
});
Object.defineProperty(exports, "NotFound", {
  enumerable: true,
  get: function () {
    return _Status.NotFound;
  }
});
Object.defineProperty(exports, "Redirect", {
  enumerable: true,
  get: function () {
    return _Redirect.Redirect;
  }
});
Object.defineProperty(exports, "Router", {
  enumerable: true,
  get: function () {
    return _ServerRouter.ServerRouter;
  }
});
Object.defineProperty(exports, "Route", {
  enumerable: true,
  get: function () {
    return _Route.Route;
  }
});
exports.useParams = exports.useLocation = exports.useRouteMatch = exports.useHistory = exports.withRouter = exports.matchPath = exports.Switch = exports.Prompt = exports.NavLink = exports.MemoryRouter = exports.Link = exports.HashRouter = exports.BrowserRouter = void 0;

var _reactRouterDom = require("react-router-dom");

var _Status = require("./modules/Status");

var _Redirect = require("./modules/Redirect");

var _ServerRouter = require("./modules/ServerRouter");

var _Route = require("./modules/Route");

/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */

/**
 * Cast each of these imports from react-router-dom to a copied-version of their
 * types.  This is necessary as the libdef defined types will not be accessible to
 * consumers of this package.
 */
const BrowserRouter = _reactRouterDom.BrowserRouter;
exports.BrowserRouter = BrowserRouter;
const HashRouter = _reactRouterDom.HashRouter;
exports.HashRouter = HashRouter;
const Link = _reactRouterDom.Link;
exports.Link = Link;
const MemoryRouter = _reactRouterDom.MemoryRouter;
exports.MemoryRouter = MemoryRouter;
const NavLink = _reactRouterDom.NavLink;
exports.NavLink = NavLink;
const Prompt = _reactRouterDom.Prompt;
exports.Prompt = Prompt;
const Switch = _reactRouterDom.Switch;
exports.Switch = Switch;
const matchPath = _reactRouterDom.matchPath;
exports.matchPath = matchPath;
const withRouter = _reactRouterDom.withRouter;
exports.withRouter = withRouter;
const useHistory = _reactRouterDom.useHistory;
exports.useHistory = useHistory;
const useRouteMatch = _reactRouterDom.useRouteMatch;
exports.useRouteMatch = useRouteMatch;
const useLocation = _reactRouterDom.useLocation;
exports.useLocation = useLocation;
const useParams = _reactRouterDom.useParams;
exports.useParams = useParams;
