// fusion-plugin-styletron-react/dist-node-cjs/browser.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireDefault(require("react"));
var _jsxRuntime = require("react/jsx-runtime");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

/* eslint-env browser */
/* global module */

let debugEngine;
let engine;
function getPlugin(getStyletronEngine) {
  const plugin = false && createPlugin({
    deps: {
      prefix: AtomicPrefixToken.optional
    },
    middleware: ({
      prefix
    }) => (ctx, next) => {
      if (ctx.element) {
        if (!engine) {
          const config = {
            hydrate: document.getElementsByClassName('_styletron_hydrate_')
          };
          if (prefix !== void 0) {
            config.prefix = prefix;
          }
          engine = getStyletronEngine(config);
        }
        if (process.env.NODE_ENV !== "production" && !debugEngine && typeof Worker !== 'undefined') {
          const worker = new Worker(workerRoute);
          worker.postMessage({
            id: 'init_wasm',
            url: wasmRoute
          });
          worker.postMessage({
            id: 'set_render_interval',
            interval: 180
          });
          if (module.hot) {
            module.hot.addStatusHandler(status => {
              if (status === 'dispose') {
                worker.postMessage({
                  id: 'invalidate'
                });
              }
            });
          }
          debugEngine = new DebugEngine(worker);
        }
        ctx.element = /*#__PURE__*/(0, _jsxRuntime.jsx)(StyletronProvider, {
          value: engine,
          debug: debugEngine,
          debugAfterHydration: Boolean(debugEngine),
          children: ctx.element
        });
      }
      return next();
    }
  });
  return plugin;
}
var _default = getPlugin;
exports.default = _default;
