// fusion-plugin-styletron-react/dist-node-cjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "AtomicPrefixToken", {
  enumerable: true,
  get: function () {
    return _constants.AtomicPrefixToken;
  }
});
exports.default = exports.StyletronMonolithicPlugin = void 0;
Object.defineProperty(exports, "styled", {
  enumerable: true,
  get: function () {
    return _styletronReact.styled;
  }
});
Object.defineProperty(exports, "withStyle", {
  enumerable: true,
  get: function () {
    return _styletronReact.withStyle;
  }
});
Object.defineProperty(exports, "withStyleDeep", {
  enumerable: true,
  get: function () {
    return _styletronReact.withStyleDeep;
  }
});
Object.defineProperty(exports, "withTransform", {
  enumerable: true,
  get: function () {
    return _styletronReact.withTransform;
  }
});
Object.defineProperty(exports, "withWrapper", {
  enumerable: true,
  get: function () {
    return _styletronReact.withWrapper;
  }
});
var _styletronReact = require("styletron-react");
var _constants = require("./constants");
var _serverAtomic = _interopRequireDefault(require("./server-atomic"));
var _serverMonolithic = _interopRequireDefault(require("./server-monolithic"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

const StyletronMonolithicPlugin = true ? _serverMonolithic.default : browserMonolithic;
exports.StyletronMonolithicPlugin = StyletronMonolithicPlugin;
var _default = true ? _serverAtomic.default : browserAtomic;
exports.default = _default;
