// fusion-core/dist-browser-cjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "CriticalChunkIdsToken", {
  enumerable: true,
  get: function () {
    return _tokens.CriticalChunkIdsToken;
  }
});
Object.defineProperty(exports, "ElementToken", {
  enumerable: true,
  get: function () {
    return _tokens.ElementToken;
  }
});
Object.defineProperty(exports, "EnableMiddlewareTimingToken", {
  enumerable: true,
  get: function () {
    return _tokens.EnableMiddlewareTimingToken;
  }
});
Object.defineProperty(exports, "ErrorHandlerToken", {
  enumerable: true,
  get: function () {
    return _tokens.ErrorHandlerToken;
  }
});
Object.defineProperty(exports, "HttpServerToken", {
  enumerable: true,
  get: function () {
    return _tokens.HttpServerToken;
  }
});
Object.defineProperty(exports, "RenderToken", {
  enumerable: true,
  get: function () {
    return _tokens.RenderToken;
  }
});
Object.defineProperty(exports, "RoutePrefixToken", {
  enumerable: true,
  get: function () {
    return _tokens.RoutePrefixToken;
  }
});
Object.defineProperty(exports, "RouteTagsToken", {
  enumerable: true,
  get: function () {
    return _tokens.RouteTagsToken;
  }
});
Object.defineProperty(exports, "SSRBodyTemplateToken", {
  enumerable: true,
  get: function () {
    return _tokens.SSRBodyTemplateToken;
  }
});
Object.defineProperty(exports, "SSRDeciderToken", {
  enumerable: true,
  get: function () {
    return _tokens.SSRDeciderToken;
  }
});
Object.defineProperty(exports, "SSRShellTemplateToken", {
  enumerable: true,
  get: function () {
    return _tokens.SSRShellTemplateToken;
  }
});
Object.defineProperty(exports, "assetUrl", {
  enumerable: true,
  get: function () {
    return _index.assetUrl;
  }
});
Object.defineProperty(exports, "chunkId", {
  enumerable: true,
  get: function () {
    return _index.chunkId;
  }
});
Object.defineProperty(exports, "compose", {
  enumerable: true,
  get: function () {
    return _compose.compose;
  }
});
Object.defineProperty(exports, "consumeSanitizedHTML", {
  enumerable: true,
  get: function () {
    return _sanitization.consumeSanitizedHTML;
  }
});
Object.defineProperty(exports, "createPlugin", {
  enumerable: true,
  get: function () {
    return _createPlugin.createPlugin;
  }
});
Object.defineProperty(exports, "createToken", {
  enumerable: true,
  get: function () {
    return _createToken.createToken;
  }
});
Object.defineProperty(exports, "dangerouslySetHTML", {
  enumerable: true,
  get: function () {
    return _sanitization.dangerouslySetHTML;
  }
});
exports.default = void 0;
Object.defineProperty(exports, "escape", {
  enumerable: true,
  get: function () {
    return _sanitization.escape;
  }
});
Object.defineProperty(exports, "getEnv", {
  enumerable: true,
  get: function () {
    return _getEnv.default;
  }
});
Object.defineProperty(exports, "html", {
  enumerable: true,
  get: function () {
    return _sanitization.html;
  }
});
Object.defineProperty(exports, "memoize", {
  enumerable: true,
  get: function () {
    return _memoize.memoize;
  }
});
Object.defineProperty(exports, "syncChunkIds", {
  enumerable: true,
  get: function () {
    return _index.syncChunkIds;
  }
});
Object.defineProperty(exports, "syncChunkPaths", {
  enumerable: true,
  get: function () {
    return _index.syncChunkPaths;
  }
});
Object.defineProperty(exports, "unescape", {
  enumerable: true,
  get: function () {
    return _sanitization.unescape;
  }
});
Object.defineProperty(exports, "unstable_EnableServerStreamingToken", {
  enumerable: true,
  get: function () {
    return _tokens.unstable_EnableServerStreamingToken;
  }
});
Object.defineProperty(exports, "unstable_withPrepareEffect", {
  enumerable: true,
  get: function () {
    return _core.unstable_withPrepareEffect;
  }
});
Object.defineProperty(exports, "withEndpoint", {
  enumerable: true,
  get: function () {
    return _core.withEndpoint;
  }
});
Object.defineProperty(exports, "withMiddleware", {
  enumerable: true,
  get: function () {
    return _core.withMiddleware;
  }
});
Object.defineProperty(exports, "withRenderSetup", {
  enumerable: true,
  get: function () {
    return _core.withRenderSetup;
  }
});
Object.defineProperty(exports, "withUniversalMiddleware", {
  enumerable: true,
  get: function () {
    return _core.withUniversalMiddleware;
  }
});
Object.defineProperty(exports, "withUniversalValue", {
  enumerable: true,
  get: function () {
    return _core.withUniversalValue;
  }
});
Object.defineProperty(exports, "workerUrl", {
  enumerable: true,
  get: function () {
    return _index.workerUrl;
  }
});
var _baseApp = _interopRequireDefault(require("./base-app"));
var _clientApp = _interopRequireDefault(require("./client-app"));
var _getEnv = _interopRequireDefault(require("./get-env"));
var _compose = require("./compose");
var _memoize = require("./memoize");
var _sanitization = require("./sanitization");
var _index = require("./virtual/index");
var _tokens = require("./tokens");
var _createPlugin = require("./create-plugin");
var _createToken = require("./create-token");
var _core = require("./core");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

const FusionApp = true ? (0, _clientApp.default)() : serverApp();
var _default = FusionApp;
exports.default = _default;
