// fusion-core/dist-browser-cjs/tokens.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.unstable_EnableServerStreamingToken = exports.SSRShellTemplateToken = exports.SSRDeciderToken = exports.SSRBodyTemplateToken = exports.RouteTagsToken = exports.RoutePrefixToken = exports.RenderToken = exports.HttpServerToken = exports.ErrorHandlerToken = exports.EnableMiddlewareTimingToken = exports.ElementToken = exports.CriticalChunkIdsToken = void 0;
var _createToken = require("./create-token");
/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

const RouteTagsToken = (0, _createToken.createToken)('RouteTagsToken');
exports.RouteTagsToken = RouteTagsToken;
const RenderToken = (0, _createToken.createToken)('RenderToken');
exports.RenderToken = RenderToken;
const ElementToken = (0, _createToken.createToken)('ElementToken');
exports.ElementToken = ElementToken;
const SSRDeciderToken = (0, _createToken.createToken)('SSRDeciderToken');
exports.SSRDeciderToken = SSRDeciderToken;
const HttpServerToken = (0, _createToken.createToken)('HttpServerToken');
exports.HttpServerToken = HttpServerToken;
const SSRBodyTemplateToken = (0, _createToken.createToken)('SSRBodyTemplateToken');
exports.SSRBodyTemplateToken = SSRBodyTemplateToken;
const SSRShellTemplateToken = (0, _createToken.createToken)('SSRShellTemplateToken');
exports.SSRShellTemplateToken = SSRShellTemplateToken;
const RoutePrefixToken = (0, _createToken.createToken)('RoutePrefixToken');
exports.RoutePrefixToken = RoutePrefixToken;
const unstable_EnableServerStreamingToken = (0, _createToken.createToken)('unstable_EnableServerStreamingToken');
exports.unstable_EnableServerStreamingToken = unstable_EnableServerStreamingToken;
const CriticalChunkIdsToken = (0, _createToken.createToken)('CriticalChunkIdsToken');
exports.CriticalChunkIdsToken = CriticalChunkIdsToken;
const EnableMiddlewareTimingToken = (0, _createToken.createToken)('EnableMiddlewareTimingToken');
exports.EnableMiddlewareTimingToken = EnableMiddlewareTimingToken;
const ErrorHandlerToken = (0, _createToken.createToken)('ErrorHandlerToken');
exports.ErrorHandlerToken = ErrorHandlerToken;
