// fusion-plugin-react-router/dist-browser-esm/modules/Redirect.js
function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */

/* eslint-disable import/no-extraneous-dependencies */
import * as React from 'react';
import PropTypes from 'prop-types';
import { Redirect as RedirectComponent } from 'react-router-dom'; // `react-router` installed by `react-router-dom`
// $FlowFixMe

import { __RouterContext as RouterContext } from 'react-router';
import { jsx as _jsx } from "react/jsx-runtime";
export class Redirect extends React.Component {
  render() {
    return /*#__PURE__*/_jsx(RouterContext.Consumer, {
      children: context => {
        const staticContext = this.context.router && this.context.router.staticContext;

        if (false && staticContext) {
          staticContext.status = parseInt(this.props.code, 10);
        }

        const {
          children,
          code,
          ...rest
        } = this.props;
        return /*#__PURE__*/_jsx(RedirectComponent, { ...rest
        });
      }
    });
  }

}

_defineProperty(Redirect, "defaultProps", {
  push: false,
  code: 307
});

Redirect.contextTypes = {
  router: PropTypes.shape({
    staticContext: PropTypes.object
  })
}; // Sanity type checking

Redirect;
