// fusion-plugin-react-router/dist-browser-esm/modules/Route.js
/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */
import * as React from 'react';
import PropTypes from 'prop-types';
import { Route as ReactRouterRouteUntyped } from 'react-router-dom';
import { jsx as _jsx } from "react/jsx-runtime";
const ReactRouterRoute = ReactRouterRouteUntyped;

const isEmptyChildren = children => React.Children.count(children) === 0;

function Route(props, context) {
  const {
    trackingId,
    component,
    render,
    children,
    ...remainingProps
  } = props;
  return /*#__PURE__*/_jsx(ReactRouterRoute, { ...remainingProps,
    // eslint-disable-next-line react/no-children-prop
    children: routeProps => {
      const {
        match
      } = routeProps;

      if (match && match.isExact) {
        if (typeof context.onRoute === 'function') {
          context.onRoute({
            page: match.path,
            title: trackingId || match.path,
            params: match.params
          });
        }
      }

      if (component) return match ? /*#__PURE__*/React.createElement(component, routeProps) : null;
      if (render) return match ? render(routeProps) : null;
      if (typeof children === 'function') return children(routeProps);
      if (children && !isEmptyChildren(children)) return children;
      return null;
    }
  });
}

Route.contextTypes = {
  onRoute: PropTypes.func
};
Route.displayName = 'FusionRoute';
export { Route };
