// fusion-plugin-react-router/dist-browser-esm/modules/ServerHistory.js
/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */
// $FlowFixMe
import { createPath, createLocation as defaultCreateLocation } from 'history';
import { addRoutePrefix, removeRoutePrefix } from './utils.js';

const createLocation = (path, prefix) => {
  const unprefixed = removeRoutePrefix(path, prefix);

  try {
    return defaultCreateLocation(unprefixed);
  } catch (e) {
    if (e instanceof URIError) {
      return defaultCreateLocation(typeof unprefixed === 'string' ? encodeURI(unprefixed) : { ...unprefixed,
        pathname: encodeURI(unprefixed.pathname)
      });
    } else {
      throw e;
    }
  }
};

const createPrefixedURL = (location, prefix) => {
  const prefixed = addRoutePrefix(location, prefix);

  if (typeof prefixed === 'string') {
    return prefixed;
  } else {
    return createPath(prefixed);
  }
};

const staticHandler = methodName => () => {
  throw new Error(`You cannot ${methodName} with server side <Router>`);
};

const noop = () => {};

export function createServerHistory(basename, context, location) {
  function createHref(location) {
    return createPrefixedURL(location, basename);
  }

  function push(path) {
    context.action = 'PUSH';
    context.location = createLocation(path, basename);
    const url = createPath(context.location);

    if (typeof url === 'string') {
      context.url = url;
    }
  }

  function replace(path) {
    context.action = 'REPLACE';
    context.location = createLocation(path, basename);
    const url = createPath(context.location);

    if (typeof url === 'string') {
      context.url = url;
    }
  }

  const history = {
    length: 0,
    createHref,
    action: 'POP',
    location: createLocation(location, basename),
    push,
    replace,
    go: staticHandler('go'),
    goBack: staticHandler('back'),
    goForward: staticHandler('forward'),
    listen: () => noop
  };
  return history;
}
