// fusion-plugin-react-router/dist-node-esm/browser.js
/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */
import { BrowserRouter as BrowserRouterUntyped, HashRouter as HashRouterUntyped, Link as LinkUntyped, MemoryRouter as MemoryRouterUntyped, NavLink as NavLinkUntyped, Prompt as PromptUntyped, StaticRouter as StaticRouterUntyped, Switch as SwitchUntyped, matchPath as matchPathUntyped, withRouter as withRouterUntyped, useHistory as useHistoryUntyped, useParams as useParamsUntyped, useRouteMatch as useRouteMatchUntyped, useLocation as useLocationUntyped } from 'react-router-dom';
import { Status, NotFound } from './modules/Status';
import { Redirect } from './modules/Redirect';
import { Router } from './modules/BrowserRouter';
import { Route } from './modules/Route';

/**
 * Cast each of these imports from react-router-dom to a copied-version of their
 * types.  This is necessary as the libdef defined types will not be accessible to
 * consumers of this package.
 */
const BrowserRouter = BrowserRouterUntyped;
const HashRouter = HashRouterUntyped;
const Link = LinkUntyped;
const MemoryRouter = MemoryRouterUntyped;
const NavLink = NavLinkUntyped;
const Prompt = PromptUntyped;
const StaticRouter = StaticRouterUntyped;
const Switch = SwitchUntyped;
const matchPath = matchPathUntyped;
const withRouter = withRouterUntyped;
const useHistory = useHistoryUntyped;
const useRouteMatch = useRouteMatchUntyped;
const useLocation = useLocationUntyped;
const useParams = useParamsUntyped;
export { BrowserRouter, HashRouter, Link, MemoryRouter, NavLink, Prompt, StaticRouter, Switch, matchPath, withRouter, Status, NotFound, Redirect, Router, Route, useHistory, useRouteMatch, useLocation, useParams };
