// fusion-plugin-react-router/dist-node-esm/index.js
/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */
import plugin, { RouterProviderToken, RouterToken, GetStaticContextToken } from './plugin';
import * as server from './server';
export default plugin;
export { RouterProviderToken, RouterToken, GetStaticContextToken };
export const BrowserRouter = false ? browser.BrowserRouter : server.BrowserRouter;
export const HashRouter = false ? browser.HashRouter : server.HashRouter;
export const Link = false ? browser.Link : server.Link;
export const matchPath = false ? browser.matchPath : server.matchPath;
export const MemoryRouter = false ? browser.MemoryRouter : server.MemoryRouter;
export const NavLink = false ? browser.NavLink : server.NavLink;
export const Prompt = false ? browser.Prompt : server.Prompt;
export const Route = false ? browser.Route : server.Route;
export const Router = false ? browser.Router : server.Router;
export const Switch = false ? browser.Switch : server.Switch;
export const withRouter = false ? browser.withRouter : server.withRouter;
export const NotFound = false ? browser.NotFound : server.NotFound;
export const Redirect = false ? browser.Redirect : server.Redirect;
export const Status = false ? browser.Status : server.Status;
export const useHistory = false ? browser.useHistory : server.useHistory;
export const useRouteMatch = false ? browser.useRouteMatch : server.useRouteMatch;
export const useLocation = false ? browser.useLocation : server.useLocation;
export const useParams = false ? browser.useParams : server.useParams;
