// fusion-plugin-react-router/dist-node-esm/modules/BrowserRouter.js
function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */
import * as React from 'react';
import PropTypes from 'prop-types';
import { Router as BaseRouter } from 'react-router-dom';
import compare from 'just-compare';
import { jsx as _jsx } from "react/jsx-runtime";
export { Status, NotFound } from './Status.js';
export { Redirect } from './Redirect.js';

class BrowserRouter extends React.Component {
  constructor(props, context) {
    super(props, context);
    this.lastTitle = null;
  }

  getChildContext() {
    const {
      __IS_PREPARE__
    } = this.context;
    return {
      onRoute: routeData => {
        if (!__IS_PREPARE__ && (routeData.title !== this.lastTitle || !compare(routeData.params, this.lastParams))) {
          this.lastTitle = routeData.title;
          this.lastParams = routeData.params;
          this.props.onRoute && this.props.onRoute(routeData);
        }
      }
    };
  }

  render() {
    const {
      Provider,
      history,
      basename
    } = this.props;
    if (!Provider) throw new Error('Missing Provider for Browser Router');
    return (
      /*#__PURE__*/
      // $FlowFixMe
      _jsx(Provider, {
        basename: basename,
        history: history,
        children: this.props.children
      })
    );
  }

}

_defineProperty(BrowserRouter, "defaultProps", {
  onRoute: () => {},
  Provider: BaseRouter
});

BrowserRouter.propTypes = {
  children: PropTypes.node,
  onRoute: PropTypes.func,
  history: PropTypes.object,
  Provider: PropTypes.any,
  basename: PropTypes.string
};
BrowserRouter.contextTypes = {
  __IS_PREPARE__: PropTypes.bool
};
BrowserRouter.childContextTypes = {
  onRoute: PropTypes.func.isRequired
};
const BrowserRouterTyped = BrowserRouter;
export { BrowserRouterTyped as Router };
