// fusion-plugin-react-router/dist-node-esm/modules/ServerRouter.js
function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */
import * as React from 'react';
import { Router as BaseRouter } from 'react-router-dom';
import { jsx as _jsx } from "react/jsx-runtime";

/**
 * The public top-level API for a "static" <Router>, so-called because it
 * can't actually change the current location. Instead, it just records
 * location changes in a context object. Useful mainly in testing and
 * server-rendering scenarios.
 */
class ServerRouter extends React.Component {
  getRouterStaticContext() {
    let context = this.props.context;

    if (context && typeof context.setCode === 'function') {
      console.warn('Using context.setCode is deprecated. Use a setter on the status prop instead');
      Object.defineProperty(context, 'status', {
        set: code => {
          context.setCode(code);
        },
        configurable: true
      });
    }

    if (context && typeof context.redirect === 'function') {
      console.warn('Using context.redirect is deprecated. Use a setter on the url prop instead');
      Object.defineProperty(context, 'url', {
        set: url => {
          context.redirect(url);
        },
        configurable: true
      });
    }

    return context || {};
  }

  getChildContext() {
    return {
      router: {
        staticContext: this.getRouterStaticContext()
      },
      onRoute: routeData => this.props.onRoute && this.props.onRoute(routeData)
    };
  }

  render() {
    const {
      Provider,
      history,
      basename,
      children
    } = this.props;
    if (!Provider) throw new Error('Missing Provider for Server Router');
    return (
      /*#__PURE__*/
      // $FlowFixMe
      _jsx(Provider, {
        basename: basename,
        history: history,
        staticContext: this.getRouterStaticContext(),
        children: children
      })
    );
  }

}

_defineProperty(ServerRouter, "defaultProps", {
  basename: '',
  context: {},
  Provider: BaseRouter,
  onRoute: () => {}
});

ServerRouter.childContextTypes = {
  router: () => {},
  onRoute: () => {}
};
const ServerRouterTyped = ServerRouter;
export { ServerRouterTyped as ServerRouter };
