// fusion-plugin-react-router/dist-node-cjs/modules/BrowserRouter.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Status", {
  enumerable: true,
  get: function () {
    return _Status.Status;
  }
});
Object.defineProperty(exports, "NotFound", {
  enumerable: true,
  get: function () {
    return _Status.NotFound;
  }
});
Object.defineProperty(exports, "Redirect", {
  enumerable: true,
  get: function () {
    return _Redirect.Redirect;
  }
});
exports.Router = void 0;

var React = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _reactRouterDom = require("react-router-dom");

var _justCompare = _interopRequireDefault(require("just-compare"));

var _jsxRuntime = require("react/jsx-runtime");

var _Status = require("./Status.js");

var _Redirect = require("./Redirect.js");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class BrowserRouter extends React.Component {
  constructor(props, context) {
    super(props, context);
    this.lastTitle = null;
  }

  getChildContext() {
    const {
      __IS_PREPARE__
    } = this.context;
    return {
      onRoute: routeData => {
        if (!__IS_PREPARE__ && (routeData.title !== this.lastTitle || !(0, _justCompare.default)(routeData.params, this.lastParams))) {
          this.lastTitle = routeData.title;
          this.lastParams = routeData.params;
          this.props.onRoute && this.props.onRoute(routeData);
        }
      }
    };
  }

  render() {
    const {
      Provider,
      history,
      basename
    } = this.props;
    if (!Provider) throw new Error('Missing Provider for Browser Router');
    return (
      /*#__PURE__*/
      // $FlowFixMe
      (0, _jsxRuntime.jsx)(Provider, {
        basename: basename,
        history: history,
        children: this.props.children
      })
    );
  }

}

_defineProperty(BrowserRouter, "defaultProps", {
  onRoute: () => {},
  Provider: _reactRouterDom.Router
});

BrowserRouter.propTypes = {
  children: _propTypes.default.node,
  onRoute: _propTypes.default.func,
  history: _propTypes.default.object,
  Provider: _propTypes.default.any,
  basename: _propTypes.default.string
};
BrowserRouter.contextTypes = {
  __IS_PREPARE__: _propTypes.default.bool
};
BrowserRouter.childContextTypes = {
  onRoute: _propTypes.default.func.isRequired
};
const BrowserRouterTyped = BrowserRouter;
exports.Router = BrowserRouterTyped;
