// fusion-plugin-react-router/dist-node-cjs/modules/ServerHistory.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createServerHistory = createServerHistory;

var _history = require("history");

var _utils = require("./utils.js");

/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */
// $FlowFixMe
const createLocation = (path, prefix) => {
  const unprefixed = (0, _utils.removeRoutePrefix)(path, prefix);

  try {
    return (0, _history.createLocation)(unprefixed);
  } catch (e) {
    if (e instanceof URIError) {
      return (0, _history.createLocation)(typeof unprefixed === 'string' ? encodeURI(unprefixed) : { ...unprefixed,
        pathname: encodeURI(unprefixed.pathname)
      });
    } else {
      throw e;
    }
  }
};

const createPrefixedURL = (location, prefix) => {
  const prefixed = (0, _utils.addRoutePrefix)(location, prefix);

  if (typeof prefixed === 'string') {
    return prefixed;
  } else {
    return (0, _history.createPath)(prefixed);
  }
};

const staticHandler = methodName => () => {
  throw new Error(`You cannot ${methodName} with server side <Router>`);
};

const noop = () => {};

function createServerHistory(basename, context, location) {
  function createHref(location) {
    return createPrefixedURL(location, basename);
  }

  function push(path) {
    context.action = 'PUSH';
    context.location = createLocation(path, basename);
    const url = (0, _history.createPath)(context.location);

    if (typeof url === 'string') {
      context.url = url;
    }
  }

  function replace(path) {
    context.action = 'REPLACE';
    context.location = createLocation(path, basename);
    const url = (0, _history.createPath)(context.location);

    if (typeof url === 'string') {
      context.url = url;
    }
  }

  const history = {
    length: 0,
    createHref,
    action: 'POP',
    location: createLocation(location, basename),
    push,
    replace,
    go: staticHandler('go'),
    goBack: staticHandler('back'),
    goForward: staticHandler('forward'),
    listen: () => noop
  };
  return history;
}
