// fusion-core/dist-node-cjs/create-token.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TokenType = exports.TokenImpl = void 0;
exports.createToken = createToken;
exports.getTokenRef = getTokenRef;
var _stackTrace = require("./stack-trace");
/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

const TokenType = Object.freeze({
  Required: 0,
  Optional: 1
});
exports.TokenType = TokenType;
function Ref() {}
class TokenImpl {
  constructor(name, ref) {
    this.name = name;
    this.ref = ref || new Ref();
    this.type = ref ? TokenType.Optional : TokenType.Required;
    this.stacks = [{
      type: 'token',
      stack: (0, _stackTrace.captureStackTrace)(createToken)
    }];
    if (!ref) {
      this.optional = new TokenImpl(name, this.ref);
    }
  }
}
exports.TokenImpl = TokenImpl;
function createToken(name) {
  return new TokenImpl(name);
}
function getTokenRef(token) {
  if (token instanceof TokenImpl) {
    return token.ref;
  }
  return token;
}
