// fusion-plugin-react-router/dist-node-cjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "RouterProviderToken", {
  enumerable: true,
  get: function () {
    return _plugin.RouterProviderToken;
  }
});
Object.defineProperty(exports, "RouterToken", {
  enumerable: true,
  get: function () {
    return _plugin.RouterToken;
  }
});
Object.defineProperty(exports, "GetStaticContextToken", {
  enumerable: true,
  get: function () {
    return _plugin.GetStaticContextToken;
  }
});
exports.useParams = exports.useLocation = exports.useRouteMatch = exports.useHistory = exports.Status = exports.Redirect = exports.NotFound = exports.withRouter = exports.Switch = exports.Router = exports.Route = exports.Prompt = exports.NavLink = exports.MemoryRouter = exports.matchPath = exports.Link = exports.HashRouter = exports.BrowserRouter = exports.default = void 0;

var _plugin = _interopRequireWildcard(require("./plugin"));

var server = _interopRequireWildcard(require("./server"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */
var _default = _plugin.default;
exports.default = _default;
const BrowserRouter = false ? browser.BrowserRouter : server.BrowserRouter;
exports.BrowserRouter = BrowserRouter;
const HashRouter = false ? browser.HashRouter : server.HashRouter;
exports.HashRouter = HashRouter;
const Link = false ? browser.Link : server.Link;
exports.Link = Link;
const matchPath = false ? browser.matchPath : server.matchPath;
exports.matchPath = matchPath;
const MemoryRouter = false ? browser.MemoryRouter : server.MemoryRouter;
exports.MemoryRouter = MemoryRouter;
const NavLink = false ? browser.NavLink : server.NavLink;
exports.NavLink = NavLink;
const Prompt = false ? browser.Prompt : server.Prompt;
exports.Prompt = Prompt;
const Route = false ? browser.Route : server.Route;
exports.Route = Route;
const Router = false ? browser.Router : server.Router;
exports.Router = Router;
const Switch = false ? browser.Switch : server.Switch;
exports.Switch = Switch;
const withRouter = false ? browser.withRouter : server.withRouter;
exports.withRouter = withRouter;
const NotFound = false ? browser.NotFound : server.NotFound;
exports.NotFound = NotFound;
const Redirect = false ? browser.Redirect : server.Redirect;
exports.Redirect = Redirect;
const Status = false ? browser.Status : server.Status;
exports.Status = Status;
const useHistory = false ? browser.useHistory : server.useHistory;
exports.useHistory = useHistory;
const useRouteMatch = false ? browser.useRouteMatch : server.useRouteMatch;
exports.useRouteMatch = useRouteMatch;
const useLocation = false ? browser.useLocation : server.useLocation;
exports.useLocation = useLocation;
const useParams = false ? browser.useParams : server.useParams;
exports.useParams = useParams;
