// fusion-core/dist-node-esm/index.js
/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import BaseApp from './base-app';
import serverApp from './server-app';
import getEnv from './get-env';
const FusionApp = false ? clientApp() : serverApp();
export default FusionApp;
export { compose } from './compose';
export { memoize } from './memoize';
// sanitization API
export { html, dangerouslySetHTML, consumeSanitizedHTML, escape, unescape } from './sanitization';

// Virtual modules
export { assetUrl, chunkId, syncChunkIds, syncChunkPaths, workerUrl } from './virtual/index';
export { RenderToken, ElementToken, SSRDeciderToken, HttpServerToken, SSRBodyTemplateToken, SSRShellTemplateToken, RoutePrefixToken, CriticalChunkIdsToken, RouteTagsToken, EnableMiddlewareTimingToken, unstable_EnableServerStreamingToken, ErrorHandlerToken } from './tokens';
export { createPlugin } from './create-plugin';
export { createToken } from './create-token';
export { getEnv };
export { withUniversalValue, withRenderSetup, unstable_withPrepareEffect, withEndpoint, withMiddleware, withUniversalMiddleware } from './core';
