// fusion-react/dist-browser-esm/context.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

/** Copyright (c) 2019 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */
import * as React from 'react';
export const FusionContext = React.createContext({});
export const ServiceContext = React.createContext(() => {
  throw new Error('`ServiceContext.Provider` was not found. This occurs if you are attempting to use `ServiceContext` in a non-React Fusion.js application.');
});
export function useService(token) {
  const getService = React.useContext(ServiceContext);
  const provides = getService(token);
  return provides;
}
export function ServiceConsumer({
  token,
  children
}) {
  return React.createElement(ServiceContext.Consumer, null, getService => {
    const provides = getService(token);
    return children(provides);
  });
}

function getServices(getService, deps) {
  const services = {};
  Object.keys(deps).forEach(name => {
    services[name] = getService(deps[name]);
  });
  return services;
}

const identity = i => i;

export function withServices(deps, mapServicesToProps = identity) {
  function resolve(getService) {
    const services = getServices(getService, deps);
    const serviceProps = mapServicesToProps(services);
    return serviceProps;
  }

  return Component => {
    return function WithServices(props) {
      return React.createElement(ServiceContext.Consumer, null, getService => React.createElement(Component, _extends({}, resolve(getService), props)));
    };
  };
}
