// fusion-plugin-react-router/dist-browser-cjs/modules/ServerRouter.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ServerRouter = void 0;

var React = _interopRequireWildcard(require("react"));

var _reactRouterDom = require("react-router-dom");

var _jsxRuntime = require("react/jsx-runtime");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/**
 * The public top-level API for a "static" <Router>, so-called because it
 * can't actually change the current location. Instead, it just records
 * location changes in a context object. Useful mainly in testing and
 * server-rendering scenarios.
 */
class ServerRouter extends React.Component {
  getRouterStaticContext() {
    let context = this.props.context;

    if (context && typeof context.setCode === 'function') {
      console.warn('Using context.setCode is deprecated. Use a setter on the status prop instead');
      Object.defineProperty(context, 'status', {
        set: code => {
          context.setCode(code);
        },
        configurable: true
      });
    }

    if (context && typeof context.redirect === 'function') {
      console.warn('Using context.redirect is deprecated. Use a setter on the url prop instead');
      Object.defineProperty(context, 'url', {
        set: url => {
          context.redirect(url);
        },
        configurable: true
      });
    }

    return context || {};
  }

  getChildContext() {
    return {
      router: {
        staticContext: this.getRouterStaticContext()
      },
      onRoute: routeData => this.props.onRoute && this.props.onRoute(routeData)
    };
  }

  render() {
    const {
      Provider,
      history,
      basename,
      children
    } = this.props;
    if (!Provider) throw new Error('Missing Provider for Server Router');
    return (
      /*#__PURE__*/
      // $FlowFixMe
      (0, _jsxRuntime.jsx)(Provider, {
        basename: basename,
        history: history,
        staticContext: this.getRouterStaticContext(),
        children: children
      })
    );
  }

}

_defineProperty(ServerRouter, "defaultProps", {
  basename: '',
  context: {},
  Provider: _reactRouterDom.Router,
  onRoute: () => {}
});

ServerRouter.childContextTypes = {
  router: () => {},
  onRoute: () => {}
};
const ServerRouterTyped = ServerRouter;
exports.ServerRouter = ServerRouterTyped;
